storage-targets/cds-feature-attachments-oss/src/main/java/com/sap/cds/feature/attachments/oss/handler/OSSAttachmentsServiceHandler.java

```diff
@@ -0,0 +1,162 @@
+			logger.error("No service binding found, hence the attachment service is not connected!");
+			this.osClient = new MockOSClient();
+			return;
+		}
+		ServiceBinding binding = bindingOpt.get();
+
+		final String host = (String) binding.getCredentials().get("host"); //AWS
+		final String containerUri = (String) binding.getCredentials().get("container_uri"); //Azure
+		final String base64EncodedPrivateKeyData = (String) binding.getCredentials().get("base64EncodedPrivateKeyData"); //GCP
+
+		// In the follwing, we check the service binding credentials to determine which client to use.
+		// We do *not* throw exceptions here, as we want to provide a fallback to the MockOSClient if no valid service binding is found.
+		// Then the rest of the application still works, but without actual attachment storage.
+		if (host != null && java.util.stream.Stream.of("aws", "s3", "amazon").anyMatch(s -> host.contains(s))) {
+			this.osClient = new AWSClient(binding, executor);
+		} else if (containerUri != null && java.util.stream.Stream.of("azure", "windows").anyMatch(s -> containerUri.contains(s))) {
+			this.osClient = new AzureClient(binding, executor);
+		} else if (base64EncodedPrivateKeyData != null) {
+		    String decoded = "";
+			try {
+				decoded = new String(java.util.Base64.getDecoder().decode(base64EncodedPrivateKeyData),java.nio.charset.StandardCharsets.UTF_8);
+			} catch (IllegalArgumentException e) {
+				logger.error("No valid base64EncodedPrivateKeyData found in Google service binding {}, hence the attachment service is not connected!", binding);
+			}
+			// Redeclaring is needed here to make the variable effectively final for the lambda expression
+			final String dec = decoded;
+			if (java.util.stream.Stream.of("google", "gcp").anyMatch(s -> dec.contains(s))) {
+				this.osClient = new GoogleClient(binding, executor);
+			} else {
+				logger.error("No valid Google service binding found in binding {}, hence the attachment service is not connected!", binding);
+				this.osClient = new MockOSClient();
+			}
+		} else {
+			logger.error("No valid service found in binding {}, hence the attachment service is not connected!", binding);
+			this.osClient = new MockOSClient();
+		}
+	}
+
+	@On
+	void createAttachment(AttachmentCreateEventContext context) throws InterruptedException, ExecutionException {
+		String fileName = context.getAttachmentEntity().getQualifiedName();
+		logger.info("OS Attachment Service handler called for creating attachment for entity {}", fileName);
+		
+		String contentId = (String) context.getAttachmentIds().get(Attachments.ID);
+		MediaData data = context.getData();
+
+		try {
+			osClient.uploadContent(data.getContent(), contentId, data.getMimeType()).get();
+			logger.info("Uploaded file {}", fileName);
+			context.setIsInternalStored(false);
+			context.setContentId(contentId);
+			context.setCompleted();
+		} catch (ObjectStoreServiceException ex) {
+			logger.error("Failed to upload file {}", fileName, ex);
```

Comment:
Currently not sure, if we need explicit logging here. I believe we always log when ServiceExceptions are thrown... Maybe you can quickly test that?